SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m

```diff
@@ -316,6 +327,13 @@ - (void)drainQueue
     dispatch_async(self.serialQueue, ^{
         if (!AppReadiness.isAppReady) {
             // We don't want to process incoming messages until storage is ready.
+            static dispatch_once_t onceToken;
+            dispatch_once(&onceToken, ^{
```

Comment:
I'll just leave this comment only once, but I feel the same about all the places in this PR where this code has been copy pasted.

Is this required for fixing the bug discussed in the PR, or is this just being extra careful?

Seeing this code duplicated in so many classes makes me feel like we have the wrong model. 

Rather than having all these classes know about "app readiness" and react to "app readiness", it seems like an "app readiness coordinator" should be telling these classes when to act. 

IMO AppDelegate is a good candidate for this "app readiness state coordinator" role (and is mostly already doing it):

e.g. 
- The AppDelegate doesn't start jobs until App is ready
- The AppDelegate doesn't act upon notifications until the app is ready
- etc.

The way I'm thinking about it - this block of code should never execute if we've initiated our dependencies coherently. In other words - if we are trying to `drainQueue` before the app is ready, that seems like a bug in whoever told the queue to start draining.

What I'd ultimately like to see is something like`OWSAssert(AppReadiness.isAppReady)`, and make any requisite changes to make that a reality, but in the interest of avoiding any big changes, would you be willing to add an `OWSFail` in all of these blocks as written?

Or are you think about this differently?

Reply:
As discussed offline, I hope we can move "app readiness coordination" into AppSetup in v2.21.

Reply:
I've made a number of changes in this direction in this PR.  Let's pursue in v2.21.